### ISSUE TITLE
[Question] Complex strategies for Learning Rate Schedule

### ISSUE THREAD

**Issue by fontanafederico** [DLR-RM/stable-baselines3#856]
### Question

How to set a learning rate schedule based on metrics derived from the policy? For example in PPO: learning_rate = 0.0003 * th.exp(self.policy.log_std).mean().item()

### Additional context

**Contaxt**. Learning rate decay allows to prevent stepping away from an optimal policy once training has converged. The [documentation](https://stable-baselines3.readthedocs.io/en/master/guide/examples.html#learning-rate-schedule) provides examples for adjusting the learning rate based on progress, which requires a prior estimate of how many steps it will take to converge. One can develop an intuition abour how long it will take to converge after familiarising with an env, conditioned to a set of hyperparameters. However, when doing hyperparameters tuning a given schedule might not be suitable across the hyperparameter space under exploration.

**Example of a use case**. Let's take PPO. Exploration is mostly driven by std assuming no entropy bonus. It makes intuitive sense to adjust the learning rate as a function of std. At the beginning of training when std is high (i.e. close to 1.) and the agent is therefore exploring, it is sensible to make larger gradient updates. During training, std will decay and approach 0 as the agent gets more confident with the policy and explores less/exploits more. 

### Checklist

- [v] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [v] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by araffin** [comment]
> How to set a learning rate schedule based on metrics derived from the policy?

You can take a look at callbacks in the doc for that ;) Callbacks have access to the model, policy, loggers and other variables.
For something even more custom, you can use a fork of SB3.

**Comment by fontanafederico** [comment]
> > How to set a learning rate schedule based on metrics derived from the policy?
> 
> You can take a look at callbacks in the doc for that ;) Callbacks have access to the model, policy, loggers and other variables. For something even more custom, you can use a fork of SB3.

That's helpful. I wasn't initially comfortable with changing the learning rate outside model initialisation due to potential spillover effects. For the benefit of others, I'm posting one way to implement a solution that works.


```
from stable_baselines3.common.callbacks import BaseCallback
import torch as th


class CallbackHyperparamsSchedulePPO(BaseCallback):
    """Learning rate = Initial learning rate * training/std"""
    
    def __init__(self):
        super().__init__()
        self._learning_rate_start = None

    def _on_rollout_end(self) -> None:
        if self._learning_rate_start is None:
            self._learning_rate_start = self.model.learning_rate
        std = th.exp(self.model.policy.log_std).mean().item()
        self.model.learning_rate = self._learning_rate_start * std
        self.model._setup_lr_schedule()
```

![image](https://user-images.githubusercontent.com/100283805/162941474-b5b83059-b6b5-4fdd-abd6-e08643164d41.png)

**Comment by araffin** [comment]
Thanks for posting the example, you could also directly change `lr_schedule` instead of calling `self.model._setup_lr_schedule()`

See https://github.com/DLR-RM/stable-baselines3/issues/735#issuecomment-1047638011
